Impelement secrets detection for Az modules

vidai-msft opened the pull request:
This pull request introduces changes primarily focused on adding a new feature for detecting and warning about potential secrets in the output of Azure PowerShell commands. The changes span across multiple files, with the addition of a new `Sanitizer` class and modifications to existing classes to incorporate the use of this new class. 

Here are the most important changes:

New feature for secrets detection:

* `src/Authentication.Abstractions/Models/ConfigKeysForCommon.cs`: Added a new constant `ShowSecretsWarning` to the `ConfigKeysForCommon` class. This constant will be used to control whether the new secrets detection feature is enabled. It is disabled by default for now.

* `src/Common/AzurePSCmdlet.cs`: Added a new `Sanitizer` object and methods for detecting secrets in the output of Azure PowerShell commands. The `Sanitizer` object is used in the `WriteObject` methods to sanitize the output before it's written to the pipeline. If any secrets are detected, a warning message is written to the console in the `WriteShowSecretsWarningMessage` method.

* `src/Common/MetricHelper.cs`: Modified the `PopulatePropertiesFromQos` method and the `AzurePSQoSEvent` class to include information about detected secrets in the telemetry data.

* `src/Common/Properties/Resources.Designer.cs` and `src/Common/Properties/Resources.resx`: Added a new localized string `ShowSecretsWarningMessage` for the warning message displayed when secrets are detected in the output.

* `src/Common/Sanitizer/AzurePSSanitizer.cs`, `src/Common/Sanitizer/DefaultProviderResolver.cs`, `src/Common/Sanitizer/DefaultSanitizerSettings.cs`: Added new classes for the `Sanitizer` object used for detecting secrets in the output. The `AzurePSSanitizer` class contains the main logic for sanitizing objects and detecting secrets. The `DefaultProviderResolver` and `DefaultSanitizerSettings` classes provide additional support for the sanitization process.

* The main logic to detect secrets in the output object is to traverse each property of this object with pre-defined types, which are `SanitizerCollectionProvider`, `SanitizerDictionaryProvider`, `SanitizerJsonArrayProvider`, `SanitizerJsonObjectProvider`, `SanitizerStringProvider`. If later on, new types are necessary, new providers as classes will be added and no existing class modification is needed.

isra-fel reviewed (commented):
1. Overall, we should minimize public types in common repo, because more public stuff means more dependencies, and more dependencies makes it difficult to maintain.
2. Instead of instanciate AzurePSSanitizer for every cmdlet, consider designing it as a singleton component like IConfigManager. This is for performance.

vidai-msft commented:
> 1. Overall, we should minimize public types in common repo, because more public stuff means more dependencies, and more dependencies makes it difficult to maintain.
> 2. Instead of instanciate AzurePSSanitizer for every cmdlet, consider designing it as a singleton component like IConfigManager. This is for performance.

Following changes were made.
1. ISanitizerSettings was renamed to ISanitizerService and the methods were also updated per discussion.
2. Moved DefaultSanitizerService and AzurePSSanitizer from out of common to Az.Accounts.

vidai-msft commented on src/Common/AzurePSCmdlet.cs:
```diff
@@ -501,12 +535,22 @@ protected new void ThrowTerminatingError(ErrorRecord errorRecord)
             FlushDebugMessages();
+            if (Sanitizer != null && Sanitizer.RequireSecretsDetection)
+            {
+                Sanitizer.Sanitize(sendToPipeline, out _sanitizerInfo);
```
Per discussion, `out` will remain there but only returns the current sanitize batch due to piping scenario.

vidai-msft commented on src/Common/Sanitizer/SanitizerProvider.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Microsoft.WindowsAzure.Commands.Common.Sanitizer
+{
+    public abstract class SanitizerProvider
```
This class has been moved out.

isra-fel reviewed (approved):
LGTM
